content/en/docs/tasks/extend-kubectl/kubectl-plugins.md

```diff
@@ -10,131 +11,266 @@ content_template: templates/task
+$ KUBECONFIG=/etc/kube/config kubectl foo config
+/etc/kube/config
+```
 
-For example, the directory structure for the `targaryen` plugin could look like this:
+Additionally, the first argument that is passed to a plugin will always be the full path to the location where it was invoked (`$0` would equal `/usr/local/bin/kubectl-foo` in our example above).
 
+### Naming a plugin
+
+As seen in the example above, a plugin determines the command path that it will implement based on its filename. Every sub-command in the command path that a plugin targets, is separated by a dash (`-`).
+For example, a plugin that wishes to be invoked whenever the command `kubectl foo bar baz` is invoked by the user, would have the filename of `kubectl-foo-bar-baz`.
+
+#### Flags and argument handling
+
+Taking our `kubectl-foo-bar-baz` plugin from the above scenario, we further explore additional cases where users invoke our plugin while providing additional flags and arguments.
+For example, in a situation where a user invokes the command `kubectl foo bar baz arg1 --flag=value arg2`, the plugin mechanism will first try to find the plugin with the longest possible name, which in this case
+would be `kubectk-foo-bar-baz-arg1`. Upon not finding that plugin, it then treats the last dash-separated value as an argument (`arg1` in this case), and attempts to find the next longest possible name, `kubectl-foo-bar-baz`.
+Upon finding a plugin with this name, it then invokes that plugin, passing all args and flags after its name to the plugin executable.
+
+Example:
+
+```bash
+# create a plugin
+$ echo '#!/bin/bash\n\necho "My first command-line argument was $1"' > kubectl-foo-bar-baz
+$ sudo chmod +x ./kubectl-foo-bar-baz
+
+# "install" our plugin by placing it on our PATH
+$ sudo mv ./kubectl-foo-bar-baz /usr/local/bin
+
+# ensure our plugin is recognized by kubectl
+$ kubectl plugin list
+The following kubectl-compatible plugins are available:
+
+/usr/local/bin/kubectl-foo-bar-baz
+
+# test that calling our plugin via a "kubectl" command works
+# even when additional arguments and flags are passed to our
+# plugin executable by the user.
+$ kubectl foo bar baz arg1 --meaningless-flag=true
+My first command-line argument was arg1
 ```
-~/.kube/plugins/
-└── targaryen
-    ├── plugin.yaml
-    └── dracarys
+
+As you can see, our plugin was found based on the `kubectl` command specified by a user, and all extra arguments and flags were passed as-is to the plugin executable once it was found.
+
+#### Names with dashes and underscores
+
+Although the `kubectl` plugin mechanism uses the dashes (`-`) in plugin filenames to determine the sequence of sub-commands that should invoke them, it is still possible to create a plugin
+command containing dashes in its commandline invocation by using underscores `_` in its filename.
```

Comment:
nit: to be consistent, use (`_`)
nit, how does this read to you: uses the dash (`-`) in plugin filenames to separate the sequence of sub-commands processed by the plugin, it is still possible to create ...